packages/core/src/state-management/repository.ts

```diff
@@ -0,0 +1,85 @@
+import { Document } from '../document';
+import DocID from '@ceramicnetwork/docid';
+import { AsyncLruMap } from './async-lru-map';
+import { StateStore } from '../store/state-store';
+import { DocumentFactory } from './document-factory';
+import { DocState } from '@ceramicnetwork/common';
+
+export class Repository {
+  readonly #map: AsyncLruMap<Document>;
+  #documentFactory?: DocumentFactory;
+  #stateStore?: StateStore
+
+  constructor(limit: number) {
+    this.#map = new AsyncLruMap(limit, async (entry) => {
+      await entry.value.close();
+    });
+  }
+
+  // This should not be here, but here we are with the initialization process.
+  setDocumentFactory(documentFactory: DocumentFactory): void {
+    this.#documentFactory = documentFactory;
+  }
+
+  // This should not be here, but here we are with the initialization process.
+  setStateStore(stateStore: StateStore) {
+    this.#stateStore = stateStore
+  }
+
+  /**
+   * Return a document, either from cache or re-constructed from state store.
+   * Adds the document to cache.
+   */
+  async get(docId: DocID): Promise<Document | undefined> {
+    const fromMemory = await this.#map.get(docId.toString());
+    if (fromMemory) {
+      return fromMemory;
+    } else {
+      if (this.#stateStore && this.#documentFactory) {
+        const docState = await this.#stateStore.load(docId);
+        if (docState) {
+          const document = await this.#documentFactory.build(docState);
+          await this.#map.set(docId.toString(), document)
+          return document
+        } else {
+          return undefined;
+        }
+      }
+    }
+  }
+
+  /**
+   * Return a document state, either from cache or from state store.
+   */
+  async docState(docId: DocID): Promise<DocState | undefined> {
+    const fromMemory = await this.#map.get(docId.toString());
+    if (fromMemory) {
+      return fromMemory.state;
+    } else {
+      if (this.#stateStore && this.#documentFactory) {
+        return this.#stateStore.load(docId);
+      }
+    }
+  }
+
+  /**
+   * Stub for adding the document.
+   */
+  async add(document: Document): Promise<void> {
```

Comment:
Since this and `delete` below don't remove the document from the state store, but just from the in-memory cache, should they be called `addToCache` and `removeFromCache`?